### PR TITLE
ENH: added a new plugin hook for counting kmers and implemented method on SequenceCollection

### PR DESCRIPTION
**Pull request by GavinHuttley**
## Summary by Sourcery

Add a plugin hook for k-mer counting and integrate it into SequenceCollection and Sequence classes with a built-in fallback implementation

New Features:
- Introduce get_count_kmers_hook in the plugin system to load third-party k-mer counting apps
- Implement count_kmers methods on SequenceCollection and Sequence that invoke a plugin or fall back to the built-in counter

Tests:
- Add tests for SequenceCollection count_kmers consistency with per-sequence counts
- Add tests for invalid hook name errors and default fallback behavior when use_hook is None

### CODE REVIEW

File: src/cogent3/_plugin.py [hunk 1]
```diff
@@ -55,6 +58,60 @@ def get_quick_tree_hook(
...
+    if name != "cogent3":
+        for extension in mgr.extensions:
+            if name and extension.module_name.startswith(name):
+                return extension.plugin(**kwargs)
+            # return first one if no name specified
+            return extension.plugin(**kwargs)
+
+        if name:
+            msg = f"Could not find count_kmers plugin for {name!r}"
+            raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply src/cogent3/_plugin.py:109]: this cannot happen because one project would need to register two separate plugins for the exact same hook in their pyproject.toml.

File: src/cogent3/core/alignment.py [hunk 2]
```diff
@@ -2495,6 +2495,57 @@ def drop_duplicated_seqs(self) -> typing_extensions.Self:
...
+        of the kmer alphabet.
+        """
+        from cogent3._plugin import get_count_kmers_hook
+
+        kwargs = {"k": k, **kwargs}
+
+        app = get_count_kmers_hook(name=use_hook, **kwargs)
+        if app is not None:
+            # we call main method directly so errors are raised here
+            return typing.cast("NumpyIntArrayType", app.main(list(self.seqs)))
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply src/cogent3/core/alignment.py:2538]: this assumption stems from the requirement that these apps are defined using the cogent3 define_app decorator, which creates that method.

File: src/cogent3/core/sequence.py [hunk 3]
```diff
@@ -441,15 +441,43 @@ def count_ambiguous(self) -> int:
...
+        self.moltype.alphabet.get_kmer_alphabet(k). See the documentation
+        for details of any third-party hooks.
         """
+        from cogent3._plugin import get_count_kmers_hook
+
+        kwargs = {"k": k, **kwargs}
+
+        app = get_count_kmers_hook(name=use_hook, **kwargs)
+        if app is not None:
+            result = cast("NumpyIntArrayType", app([self]))
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply src/cogent3/core/sequence.py:477]: that's the third-party developers problem.
